File: Esame_22/PYTHON/Uni.py
import sys
import argparse # mi serve per fare il parsing

# IMPORTO LE CLASSI
from Corso import Corso # from Modulo import Classe
from Teoria import Teoria
from  Laboratorio import  Laboratorio
from Studente import  Studente
from Esame import  Esame


# Definisco il programma dentro una CLASSE, usando la tecnica di trasformare un modulo
# in un programma mettendolo alla fine del codice con l'if
class Uni:
    #COSTRUTTORE
    def __init__(self, arg): # prende un argomento e lo memorizza in _arg
        self._arg = arg

    def main(self): # CREO IL METODO MAIN che poi richiamo in fondo
        corsi = [] # definisco i corsi e gli studenti come liste inizialmente vuote
        studenti = []
        nomeCorsi = {} # dizionario che associa i codici del corso al loro nome
        # Il dizionario serve per rispondere alla domanda 5 --> nome del corso in cui si è preso il voto + alto

        try:
            f = open("corsi.txt","rt")
            line = f.readline().strip() # readline() restituisce una linea con l'accapo
                                        # il metodo strip() rimuove l'accapo finale
            while line != "" : # Python finisce di leggere le righe quando restituisce la riga vuota
                tok = line.split() # con split posso memorizzare le parole di una riga in una lista
                codiceCorso = int(tok[0]) ##cast esplicito##
                tipoCorso = tok[1]
                line = f.readline().strip() # leggo la nuova riga gestendo l'accapo
                nomeCorso = line
                line = f.readline().strip() # leggo la nuova riga gestendo l'accapo
                docente = line

                nomeCorsi[codiceCorso] = nomeCorso # aggiungo questa associazione (codice,nome)
                                                   # al dizionario nomeCorsi --> MAPPA
                line = f.readline().strip() # leggo la nuova riga gestendo l'accapo
                if tipoCorso == "teoria":
                    tok = line.split()
                    codiceAula = tok[0]
                    oreSettimanali = tok[1]
                    orePerLezione = tok[2]
                                                # creo oggetto
                    corsoT = Teoria(codiceCorso, nomeCorso, docente, codiceAula, oreSettimanali,orePerLezione)
                    corsi.append( corsoT ) # aggiungo in LISTA
                else:
                    nomeLaboratorio = line
                    line = f.readline().strip()
                    nomeAssistente = line
                    line = f.readline().strip()
                    postazioni = int(line)

                    corsoL = Laboratorio(codiceCorso, nomeCorso, docente, nomeLaboratorio, nomeAssistente, postazioni)
                    corsi.append( corsoL )
                line = f.readline().strip() # leggo riga vuota
                line = f.readline().strip() # leggo la prossima riga per tornare al ciclo while
            f.close() # chiudo il file
        except IOError:
            print("IO Error found")
        except:
            print("Unexpected error:", sys.exc_info()[0] )
            raise # rilancia l'eccezione

        try:
            f = open("studenti.txt","rt")
            line = f.readline().strip()
            while line!= "":
                matricola = int(line)
                line = f.readline().strip()
                nomeStudente = line

                stud = Studente(matricola,nomeStudente)
                studenti.append(stud)
                line = f.readline().strip()
                while line != "" and line != "\n":
                    tok = line.split()
                    codice = int(tok[0])
                    voto = tok[1]
                    e = Esame(codice, voto)
                    stud.addEsame(e)
                    line = f.readline().strip()
                line = f.readline().strip()
            f.close() #
        except(IOError):
            print("IO Error found.")
        except:
            print("Unexpected error:", sys.exc_info()[0])

        # PUNTO 3
        print("nome, codice, docente, tipo, aula, ore sett., ore/lez., lab., assistente, postazioni")
        for c in corsi:
            print( c.toString() )

        # PUNTO 4
        for s in studenti:
            print( s.toString() )
        print("\n")

        # PUNTO 5: leggere da riga di comando il codice, e stampare nome e cognome dello studente , nome del corso
        # in cui ha preso il voto più alto e il voto
        for s in studenti:
            if( s.getMatricola() == int(self._arg) ): # ottengo la matricola dello studente e la confronto
                                                      # con l'intero presente in self._arg passato come argomento
                maxe = s.votoMax()
                codc = maxe.getCorso()
                print(s.getNome() + " " + nomeCorsi[codc] + " " + maxe.getVotoAsString() )

# Se nome = main --> sto eseguendo il codice da riga di comando
# chiamo il metodo ArgumentParser() del modulo argparse per restituire un parser
# aggiungo l'argomento matricola con l'azione di memorizzare l'argomento
# chiamo il metodo parse_args() e metto il risultato in matr
# creo un oggetto di classe Uni passandogli la matricola
# chiamo il main della classe Uni
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('matricola', action='store')
    matr = parser.parse_args()
    u = Uni(matr.matricola) # self._arg è la matricola che ho letto da tastiera e che ho passato al costruttore
                            # init che l'ha messa nella variabile arg
    u.main()
# --> Per il punto 5: vuol dire che il codice dello studente viene passato come matricola
# e viene messo dal costruttore Uni() dentro la variabile arg definita all'inizio, quindi

File: Esame_22/JAVA/Uni.java
import java.io.*;
import java.util.*;

public class Uni {
	// LISTE
	static List<Corso> corsi = new LinkedList<Corso>(); // Lista per i corsi
	static List<Studente> studenti = new LinkedList<Studente>(); // Lista per gli studenti

	// MAPPA
	static Map<Integer, String> nomeCorsi = new HashMap<Integer, String>();

	// Punto 1: memorizzazione dei corsi
	public static void main(String[] args) {

		try {
			BufferedReader br = new BufferedReader(new FileReader("corsi.txt"));
			String line = br.readLine(); // leggo la prima linea
			while (line != null) {
				StringTokenizer tok = new StringTokenizer(line); // risolve il problema degli spazi
				int codiceCorso = Integer.parseInt(tok.nextToken());
				String tipoCorso = tok.nextToken();
				line = br.readLine(); // vado a capo per leggere la prossima linea
				String nomeCorso = line;
				line = br.readLine(); // vado a capo per leggere la prossima linea
				String docente = line;

				nomeCorsi.put(codiceCorso, nomeCorso); // per punto 5:
				// con il metodo put aggiungo una coppia codice nome alla mappa nomeCorsi
				// così nella mappa ho coppie (chiave, valore) --> (codiceCorso, nomeCorso)

				line = br.readLine(); // vado a capo per leggere la prossima linea
				if (tipoCorso.equals("teoria")) {
					tok = new StringTokenizer(line);
					int codiceAula = Integer.parseInt(tok.nextToken());
					int oreSettimanali = Integer.parseInt(tok.nextToken());
					double orePerLezione = Double.parseDouble(tok.nextToken());

					Corso teoria = new Teoria(codiceCorso, nomeCorso, docente, codiceAula, oreSettimanali,
							orePerLezione);
					corsi.add(teoria); // Aggiungo questo corso alla lista dei corsi

				} else {
					// il tokenizer non serve perchè vado sempre a capo
					String nomeLaboratorio = line;
					line = br.readLine(); // vado a capo per leggere la prossima linea
					String nomeAssistente = line;
					line = br.readLine(); // vado a capo per leggere la prossima linea
					int postazioni = Integer.parseInt(line);

					Corso laboratorio = new Laboratorio(codiceCorso, nomeCorso, docente, nomeLaboratorio,
							nomeAssistente, postazioni);
					corsi.add(laboratorio); // Aggiungo questo corso alla lista dei corsi

				}
				line = br.readLine(); // Ho finito di leggere un Corso e vado a capo
				line = br.readLine(); // una riga vuota separa i dati relativi ai diversi corsi
			}
			br.close(); // chiudo il file
		} catch (IOException e) {
			System.err.println(e); // size + ctrl + invio
		} catch (Exception e) {
			System.err.println(e);
		}

		// Punto 2: Lettura di tutti gli studenti:
		try {
			BufferedReader br = new BufferedReader(new FileReader("studenti.txt"));
			String line = br.readLine();
			while (line != null) {
				int matricola = Integer.parseInt(line);
				line = br.readLine();
				String nomeCognomeStudente = line;
				Studente stud = new Studente(matricola, nomeCognomeStudente);

				studenti.add(stud); // Aggiungo qesto studente alla lista degli studenti
				line = br.readLine(); // vado a capo per leggere la prossima linea
				// Sotto forma di elenco: devo scandire questo elenco che ha lunghezza variabile
				while (line != null && !line.equals("")) {
					// null vuol dire che il puntatore non punta da nessuna parte. Non è lecita la
					// stringa
					// invece stringa vuota "" il puntatore punta a questa stringa lecita
					StringTokenizer tok = new StringTokenizer(line);
					int codiceCorso = Integer.parseInt(tok.nextToken());
					String voto = tok.nextToken();

					Esame e = new Esame(codiceCorso, voto); //
					// devo aggiungere questo esame alla lista degli studenti:
					// allora, in Studente aggiungo un metodo addEsame().
					stud.addEsame(e); // aggiungo l'esame nella lista degli studenti
					line = br.readLine(); // continuo finchè
				}
				line = br.readLine(); // vado a leggere la prossima linea
			}
			br.close(); // chiudo il file

		} catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}

		// PUNTO 3: stampare l'elenco di tutti i corsi
		System.out.println("nome, codice, docente, tipo, aula, ore sett., ore/lez., lab., assistente, postazioni\n");
		for (Corso c : corsi)
			System.out.println(c); // --> fare la toString in classe base e classi derivate

		System.out.println();

		// PUNTO 4: stampare l'elenco degli studenti
		for (Studente s : studenti)
			System.out.println(s); // --> fare la toString in classe Studente e classe dell'elenco

		System.out.println(); // salto uno spazio

		// PUNTO 5: leggere da riga di comando il codice di uno studente e stampare il
		// nome e cognome
		// dello studente, il nome del corso in cui ha preso il voto più alto e il voto

		// per leggere da linea di comando uso gli argomenti --> args[0], args[1], .....
		int matricola = Integer.parseInt(args[0]);
		for (Studente s : studenti) { // scandisco la lista degli studenti
			if (s.getMatricola() == matricola) // se matricola inserita corrisponde a quella della lista
			{
				Esame maxe = s.votoMax();
				int codc = maxe.getCorso();
				System.out
						.println(s.getNomeCognomeStudente() + " " + nomeCorsi.get(codc) + " " + maxe.getVotoAsString());

				// PER STAMPARE IL NOME DEL CORSO: ( io del corso conosco il codice, ma non
				// conosco il nome
				// potrei o fare un ciclo sull'elenco dei corsi ed andare a vedere, per ogni
				// corso se il
				// codice è uguale a quello che ho io: se è uguale prendo il nome e così via).

				// OPPURE: costruisco una mappa che va dal codice del corso al suo nome,
				// quindi una mappa da intero a stringa che si chiama nomeCorsi:
				// implementata come una hashMap: leggere come è implementata e il metodo
				// mappa.put()
				// col metodo mappa.get() ottengo il valore associato alla chiave codc
				// ossia ottengo il nome del corso
			}
		}

	} // fine main

} // fine classe
